### ISSUE TITLE
how to use this on a subset of Wikidata?

### ISSUE THREAD

**Issue by dpriskorn** [DaniFdezAlvarez/shexer#149]
would it be possible to make notebook or example for this usecase?
I found https://github.com/DaniFdezAlvarez/shexer/tree/develop/experiments/wikidata/countries_and_cities but I don't see the code that produced the result there. A notebook would be preferred so I can easily play with it and adapt it to my needs.

**Comment by DaniFdezAlvarez** [comment]
Hi @dpriskorn . You can find example code showing how to use sheXer against the Wikidata endpoint here: [https://github.com/DaniFdezAlvarez/shexer/blob/master/doc/shexer_wikidata_tutorial.ipynb]( https://github.com/DaniFdezAlvarez/shexer/blob/master/doc/shexer_wikidata_tutorial.ipynb).

If, instead of the endpoint, you want to work against a local subset of a Wikidata dump, you can also do it. But the way of doing it depends on your use case.  In case the subset is in turtle/ntriples, you can do it trivially. But if your subset is in JSON, sheXer don't offer currently a parser for that format (if you need it, add an issue and I'll try to implement it). So in that last case you should translate it first to RDF or load the graph in an rdflib graph. 

If you provide more details about your use case, I can send you ad-hoc example code for it.

**Comment by DaniFdezAlvarez** [comment]
@dpriskorn , were you able to run it for your case? do you need some help or may we close this issue?

**Comment by dpriskorn** [comment]
Thanks, I'll reopen if needed

**Comment by cjuigne** [comment]
Hi @DaniFdezAlvarez, I'm interested in using sheXer, but several of the Jupyter notebooks aren't working for me. Especially the ones using Wikidata, such as https://github.com/DaniFdezAlvarez/shexer/blob/master/doc/shexer_wikidata_tutorial.ipynb and https://github.com/weso/shexer/blob/master/doc/Example_federated_shapes.ipynb. It always returns “TypeError: ‘NoneType’ object is not subscriptable”. Has something changed, or is there something I need to do to make it work?

**Comment by DaniFdezAlvarez** [comment]
Hi @cjuigne , I suspect there is a dependence for labelling wikidata elements causing that issue, but I'm not 100% sure. I'll try to fix this asap (in the following days).

**Comment by DaniFdezAlvarez** [comment]
Hi @cjuigne,

The actual error is not being output in the console, but it seems to be related to Wikidata endpoint rate limits. When querying their service from sheXer scripts (using two different IPs), I'm consistently getting:

HTTP Error 429: Aggressively rate-limiting to 1 req / min - this rule was created during active wdqs outage (797a132).

That part of sheXer’s code has been unchanged for several years and seems to work correctly for the rest of the endpoints, so I'm not sure whether this is due to a recent change or if it is just temporary. However, there is still room for improvement in any case (see _query_endpoint_json_result in https://github.com/weso/shexer/blob/master/shexer/io/sparql/query.py).

I think, at least, a better User-Agent should be added. Also, a better retry policy and error-reporting strategy would help. I might not be able to work on this again for a week, though, but I’ll look into it.

**Comment by cjuigne** [comment]
Hi @DaniFdezAlvarez, 
Thanks for identifying the issue so quickly!
Are there any other available examples of federated shapes that use a different endpoint so I can keep testing in the meantime?

**Comment by DaniFdezAlvarez** [comment]
Hi @cjuigne ,

Maybe the easiest way to go here is to mofidy the jupyter with the federated example, as it runs 3 toy examples and one of those does not involve Wikidata [https://github.com/weso/shexer/blob/master/doc/Example_federated_shapes.ipynb](https://github.com/weso/shexer/blob/master/doc/Example_federated_shapes.ipynb).

It should work against the DBpedia endpoint if you do the following:
- In the __main__, comment everything except the line with **run_single_inverse(toy_data_example)**
- In the function run_single_inverse, set the parameter wikidata_annotation to False.

The output may be overwhleming though. As it is configured there, sheXer will locate the instances in DBpedia declared to be schema:sameAs of some entities in a local graph, and produce shapes describing the structure of the local classes (countries,cities) and shapes describing the properties observed when combining the properties of the local and the federated instances (again, countries and cities). Along with tons of annotations and statistics.

 If you already have a particular use case, I could help with the configuring sheXer.

**Comment by DaniFdezAlvarez** [comment]
@cjuigne , I'm closing this issue. I've already added the user-agent header, error report, and retry strategy that I mentioned. Still getting issues from Wikidata endpoint though, but I'm not really sure wheter I can handlt them or that's just how Wikidata works now.

Feel free to reopen it or (better) create a new one if you still think there is a bug in sheXer or figure out how to handle Wikidata's endpoint to avoid errors due to too many request. Also, in case you are interested in Wikidata content using sheXer, I can suggest using the QLever endpoint instead of the official one: https://qlever.dev/wikidata. I've tried with it and it seems that it does not reject sheXer's requests

**Comment by cjuigne** [comment]
Hi @DaniFdezAlvarez,
Sorry for the late reply.
Thank you very much for your answers. I just tested the “run_single_inverse” example without Wikidata, and it does indeed work. I’ll analyze it in detail to understand how it works and get back to you if I have any questions regarding my specific use case.

By the way, I also encountered errors when applying SheXer to another graph that also uses qlever. But I haven’t investigated the issue further (so it might be more of a qlever issue than a Wikidata issue).
